library/std/src/os/uefi/env.rs

```diff
@@ -0,0 +1,129 @@
+use r_efi::efi::{Guid, Handle};
+use r_efi::system;
+
+static GLOBAL_SYSTEM_TABLE: AtomicPtr<SystemTable> = AtomicPtr::new(crate::ptr::null_mut());
+static GLOBAL_SYSTEM_HANDLE: AtomicPtr<c_void> = AtomicPtr::new(crate::ptr::null_mut());
+
+/// Initializes Global Atomic Pointers to SystemTable and Handle.
+/// Should only be called once in the program execution under normal circumstances.
+/// The caller should ensure that the pointers are valid.
+pub(crate) fn init_globals(handle: NonNull<c_void>, system_table: NonNull<SystemTable>) {
+    GLOBAL_SYSTEM_TABLE.store(system_table.as_ptr(), Ordering::SeqCst);
+    GLOBAL_SYSTEM_HANDLE.store(handle.as_ptr(), Ordering::SeqCst);
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Get the SystemTable Pointer.
+pub fn get_system_table() -> Option<NonNull<SystemTable>> {
+    NonNull::new(GLOBAL_SYSTEM_TABLE.load(Ordering::SeqCst))
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Get the SystemHandle Pointer.
+pub fn get_system_handle() -> Option<NonNull<c_void>> {
+    NonNull::new(GLOBAL_SYSTEM_HANDLE.load(Ordering::SeqCst))
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Get the BootServices Pointer.
+pub fn get_boot_services() -> Option<NonNull<BootServices>> {
+    let system_table = get_system_table()?;
+    let boot_services = unsafe { (*system_table.as_ptr()).boot_services };
+    NonNull::new(boot_services)
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Get the RuntimeServices Pointer.
+pub fn get_runtime_services() -> Option<NonNull<RuntimeServices>> {
+    let system_table = get_system_table()?;
+    let runtime_services = unsafe { (*system_table.as_ptr()).runtime_services };
+    NonNull::new(runtime_services)
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Open Protocol on a handle
+/// Implemented using `EFI_BOOT_SERVICES.OpenProtocol()`
+pub fn open_protocol<T>(
+    handle: NonNull<c_void>,
+    mut protocol_guid: Guid,
+) -> io::Result<NonNull<T>> {
+    let boot_services = get_boot_services()
+        .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Failed to get BootServices"))?;
+    let system_handle = get_system_handle()
+        .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Failed to get System Handle"))?;
+    let mut protocol: MaybeUninit<*mut T> = MaybeUninit::uninit();
+
+    let r = unsafe {
+        ((*boot_services.as_ptr()).open_protocol)(
+            handle.as_ptr(),
+            &mut protocol_guid,
+            protocol.as_mut_ptr().cast(),
+            system_handle.as_ptr(),
+            crate::ptr::null_mut(),
+            system::OPEN_PROTOCOL_GET_PROTOCOL,
+        )
+    };
+
+    if r.is_error() {
+        Err(super::io::status_to_io_error(r))
+    } else {
+        NonNull::new(unsafe { protocol.assume_init() })
+            .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Null Protocol"))
+    }
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
```

Comment:
I think a tracking issue will need to be opened before this can be merged.

Reply:
I have started a new issue: #100499